ffi/custom_passes.cpp

```diff
@@ -920,27 +918,68 @@ struct RefPrunePass : public FunctionPass {
+            auto int_operand = dyn_cast<ConstantInt>(operand);
+            if (int_operand && int_operand->isOneValue()) {
+                return true;
+            }
+            // If the operand is a PHI node, check if there is a non-exception
+            // path. We don't know which path we're on, but since the
+            // exceptional path will lookahead, so if there is a non-exceptional
+            // path, we can assume were on it.
+            auto phi_operand = dyn_cast<PHINode>(operand);
+            if (phi_operand) {
+                for (auto& phi_arg_value : phi_operand->incoming_values()) {
+                    auto arg_value = dyn_cast<ConstantInt>(phi_arg_value);
+                    if (arg_value && !arg_value->isOneValue()) {
+                        return false;
+                    }
+                }
+                return true;
+            }
             return false;
-        // Fetch the ref to the metadata operand at location 0
-        auto &operand = md->getOperand(0);
-        // and then cast the const as Metadata (Numba sets this as literal 1)
-        auto data = dyn_cast<ConstantAsMetadata>(operand.get());
-        // If dyn_cast failed type check then return false
-        if (!data)
+        } else if (term->getOpcode() == Instruction::Br &&
+                   term->getNumOperands() == 1) {
+            // If it's a branch, it might be a common return block
+            auto first =
+                term->getSuccessor(0)->getFirstNonPHIOrDbgOrLifetime(true);
+            if (!first) {
+                // Malformed block with no terminal instruction
+                return false;
+            }
+            // Our one and only instruction should be a return
+            if (first->getOpcode() != Instruction::Ret) {
+                return false;
+            }
+            // With one operand
+            if (first->getNumOperands() != 1) {
+                return false;
+            }
+            auto operand = first->getOperand(0);
+            // If the operand is a constant, check if it indicates an exception
+            auto int_operand = dyn_cast<ConstantInt>(operand);
+            if (int_operand && int_operand->isOneValue()) {
+                return true;
+            }
+            // If the operand is a PHI node, check if the path we're on will
+            // yield a value indicating an exception
+            auto phi_operand = dyn_cast<PHINode>(operand);
+            if (phi_operand) {
+                auto arg_value = dyn_cast<ConstantInt>(
+                    phi_operand->getIncomingValueForBlock(bb));
+                return arg_value && arg_value->isOneValue();
+            }
+            // This path doesn't raise
             return false;
-        // get the value of the casted metadata and then return bool on whether
-        // it is the number one.
-        return data->getValue()->isOneValue();
+        }
+        return false;
```

Comment:
I ran the entire Numba testsuite with an asserting that number of `fanout_raise` to be zero and the testsuite passed. That means nothing is ever trigger the `fanout_raise` case any more. I also haven't found any Numba jit code that will trigger this case.

Reply:
I'll try to confirm this again tomorrow

Reply:
ah, i just found `numba::typed::typedlist::_append[abi:v3][abi:c8tJTIcFHzwl2ILiXkcBV0KBSgP9CGZpAgA_3d](ListType_5bint32_5d, int)` is triggering `fanout_raise`. i might have an older commit of this PR installed previously.